### PR TITLE
Servlet: filter API

### LINKED ISSUE #4142: Servlet: Filter API.
#### Proposed title of article
Servlet: Filter API.

#### Proposed article introduction
When an application is pre-and post-processed, the object being invoked is filters. You can use and configure filters for some filtering tasks. This is included in the Servlet API. The filter interface is found in the javax.servlet package. In order to create a filter, we have to implement the filter interface.
You can have a number of filters for pre-and post-processing requests. In the deployment descriptor of the web application, filters are configured.

Interfaces available in servlet filters, and their methods will be discussed in this article.

#### Key takeaways 
1. Introduction.
2. Filter usage and its advantages.
3. Filter interface
4. FilterChain interface
5. FilterConfig interface
6. An example showing the use of a servlet filter

#### Article quality
How is your article unique? Tell us what makes your approach different from similar articles that have published on the same topic?
Is yours more in-depth? Does it cover additional topics? Do you provide handy tips or anecdotal advice?

#### References
N/A

### PR DESCRIPTION
**Pull request by Tumorbenson**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [ ] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #4142

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by Tumorbenson** [comment]
How are you doing, @BrianaNzivu? I'd like to know how far along the review of this article is, and if there are any changes or modifications you'd like me to make.

File: content/articles/Servlet-Filter-API/index.md [hunk 1]
```diff
@@ -0,0 +1,307 @@
...
+
+  - [Filter usage](#filter-usage)
+
+  - [Advantages](#advantages)	
+
+- [Filter interface](#filter-interface)	
+- [FilterChain interface](#filterchain-interface)
+- [FilterConfig interface](#filterconfig-interface)
+- [Authentication Example using filter](#authentication-example-using-filter)
+- [Conclusion](#conclusion)
```
Comment (BrianaNzivu) [inline content/articles/Servlet-Filter-API/index.md:27]: You forgot to mention the goals/key takeaways of this article. Have it before the table of contets.

File: content/articles/Servlet-Filter-API/index.md [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+        <input type="text" name="userName" placeholder="Enter user name" required>
+        <input type="password" name="uPassword" placeholder="Enter password" required>
+        <input type="submit" value="login">
+    </form>
+</body>
+
+</html>
+```
+
+**Step 3:** Filter 1 and `com.demo` package:
```
Comment (BrianaNzivu) [inline content/articles/Servlet-Filter-API/index.md:148]: Have the step headings as h4 by that I mean (####)

**Review by BrianaNzivu — CHANGES_REQUESTED** [review]
Hey @Tumorbenson 👋  .

First of all, welcome to Section - EngEd. :tada: 
We are happy to have you here.

Good job on the article. It is a very interesting read.

Just a few changes I need you to make as we continue with the preliminary revisions:
- Your article has quite a lot of grammatical mistakes and unclear sentences. You can use 
[Grammarly](https://app.grammarly.com/ddocs/1287976180) to correct it. 
- I have noticed some of your images are large. All your images including your avatar image should be scaled down to 600 x 400 px as required. You can use [Befunky](https://www.befunky.com/) to do this.
- Please rename your png images from `PNG` to `png`.

In case of any question or help needed please do not hesitate to ping me here :smile:.
Also, notify me when you are done.

Regards.

**Comment by Kayere** [comment]
Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content. Be sure to see our resources page to see more info on plagiarism and what is considered as such.

[Servlet_ filter API #4461 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7452534/Servlet_.filter.API.4461.Copyleaks.Report.pdf)

**Comment by Tumorbenson** [comment]
> Upon running your article through our 3rd party plagiarism checker it seemed to raise a few flags and the % was higher than we typically accept. Please see attached PDF - and revisit the article to ensure we are contributing wholly unique and original content. Be sure to see our resources page to see more info on plagiarism and what is considered as such.
> 
> [Servlet_ filter API #4461 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7452534/Servlet_.filter.API.4461.Copyleaks.Report.pdf)

@Kayere  I'm done.

**Comment by Kayere** [comment]
@Tumorbenson - It is now at 16%

[(Second check) Servlet_ filter API #4461 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7459460/Second.check.Servlet_.filter.API.4461.Copyleaks.Report.pdf)

**Comment by Tumorbenson** [comment]
> @Tumorbenson - It is now at 16%
> 
> [(Second check) Servlet_ filter API #4461 Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7459460/Second.check.Servlet_.filter.API.4461.Copyleaks.Report.pdf)

Hello @Kayere? Imports of Java code and inbult methods account for the majority of the total plag. If I try to delete or edit the code, it will not fully execute.
Thank you.

File: content/articles/Servlet-Filter-API/index.md [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+
+### Authentication Example using filter
+In this example, we're demonstrating how a `Filter` is used to achieve authentication. We're using two filters here: one to verify if a password is greater than seven characters, and the other to check both the user name and password supplied by the user.
+Filter 1 will forward the request to a second Filter if the password length is correct. Otherwise, an error notice will appear.
+
+If the user name and password provided by the user are both correct, the second Filter will route the request to the Servlet page; otherwise, an error message will be displayed.
+#### Steps to create Authentication Example
+We assumed you already knew how to use the eclipse IDE and configure the Tomcat server in this example.
+
+#### Step 1:
```
Comment (BrianaNzivu) [inline content/articles/Servlet-Filter-API/index.md:142]: Please give a brief title for this steps and all the other steps i.e
Step one: Creating a file

**Review by BrianaNzivu — APPROVED** [review]
Hey @Tumorbenson   👋 .
I would love to sincerely apologize for the delayed review on your article.
The article is a very interesting read. Good job!
Just a few things I'd like to point out:
1.  It is always advisable to submit an article that is well polished. By this, I mean having an article that does not have too many grammatical corrections and functional code. I suggest that you check out our [resources](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) page which has links to various resourceful tools for writing. 
2. I'd like to recommend taking Google's technical writing [course](https://developers.google.com/tech-writing) which is absolutely free and can come in handy.
3. Lastly please go through **ALL** the corrections I have made so that you can not repeat the same mistakes.
4.  Don't  forget to name your images properly and to always have their size as 600 x 400 px.

That's it! :smile: 

A final review will be done now by a content moderator.
Otherwise, congratulations on publishing your first article🍾.

In case of any question or clarification (currently or in the future), do not hesitate to reach out via [Slack](https://app.slack.com/client/T5X14SH7G/D01A6MT673R) or my personal email (briananzivu@gmail.com). 
I am **ALWAYS** here to help.

It has been a pleasure reading and reviewing this article.

Looking forward to seeing you contribute more to EngEd. 🚀

Kind Regards.

**Comment by ahmadmardeni1** [comment]
👋 Thank you very much for contributing your 1st article - and Welcome to the EngEd Program. @Tumorbenson 

There are some significant readability challenges/grammar mistakes/unclear sentences in your article. You are not explaining to the reader well about what filters are and why they would want to use them/read about them.

We do not provide editing services and recommend that students leverage online editing services (there are both free and paid services that you can explore), or try engaging with the community or other peers in your network to have them review prior to submitting a revised proposal.

Please feel free to review our resources page on how to better improve our technical writing - and feel free to explore another topic when ready. 👍

Thank you very much for your review and I'm sure the level of feedback you provided is very helpful. @BrianaNzivu

**Comment by Tumorbenson** [comment]
> 👋 Thank you very much for contributing your 1st article - and Welcome to the EngEd Program. @Tumorbenson
> 
> There are some significant readability challenges/grammar mistakes/unclear sentences in your article. You are not explaining to the reader well about what filters are and why they would want to use them/read about them.
> 
> We do not provide editing services and recommend that students leverage online editing services (there are both free and paid services that you can explore), or try engaging with the community or other peers in your network to have them review prior to submitting a revised proposal.
> 
> Please feel free to review our resources page on how to better improve our technical writing - and feel free to explore another topic when ready. 👍
> 
> Thank you very much for your review and I'm sure the level of feedback you provided is very helpful. @BrianaNzivu

Hello @hectorkambow and @ahmadmardeni1   ? Would it be possible for me to make some changes to this article and resubmit it for review?

**Comment by ahmadmardeni1** [comment]
You can open a new PR. @ me and your peer reviewer, but please make sure your article is ready to be reviewed. @Tumorbenson 
Note for the peer reviewer: (add the duplicate label please).

**Pull request closed without merging**